bazel/utils/container/container.bzl

```diff
@@ -320,3 +532,214 @@ container_repo = rule(
+            allow_single_file = [".sh"],
+            executable = True,
+            cfg = "exec",
+            default = "@enkit//bazel/utils/container:install_ubuntu.sh",
+        ),
+        "preinstall_script": attr.label(
+            doc = "Script that executes via chroot in the bootstrap env before the install script",
+            allow_single_file = [".sh"],
+            executable = True,
+            cfg = "exec",
+            default = "@enkit//bazel/utils/container:prepostinstall_ubuntu.sh",
+        ),
+        "postinstall_script": attr.label(
+            doc = "Script that executes via chroot in the bootstrap env after the install script",
+            allow_single_file = [".sh"],
+            executable = True,
+            cfg = "exec",
+            default = "@enkit//bazel/utils/container:prepostinstall_ubuntu.sh",
+        ),
+        "exclude_pkgs": attr.string_list(
+            doc = "List of packages to not install",
+        ),
+        "tars": attr.label_list(
+            doc = "List of tarballs to extract into the debootstrap environment",
+            allow_files = [".tar", ".tar.gz"],
+        ),
+    }
+)
+
+def ubuntu_pkg_rule_impl(ctx):
+    outfile = ctx.actions.declare_file("%s.tar" % ctx.attr.name)
+    args = ctx.actions.args()
+    args.add(ctx.attr.arch)
+    args.add(",".join(ctx.attr.components))
+    args.add(ctx.attr.pkg)
+    args.add(ctx.attr.distro)
+    args.add(ctx.attr.mirror)
+    args.add(",".join(ctx.attr.exclude_pkgs))
+    args.add(outfile)
+
+    ctx.actions.run(
+        executable = ctx.executable.bootstrap_script,
+        outputs = [outfile],
+        arguments = [args],
+    )
+
+    return [
+        DefaultInfo(files = depset([outfile])),
+    ]
+
+ubuntu_pkg_rule = rule(
+    implementation = ubuntu_pkg_rule_impl,
+    attrs = bootstrap_attrs | {
+        "pkg": attr.string(
+            doc = "Name of the package to download",
+            mandatory = True,
+        ),
+        "exclude_pkgs": attr.string_list(
+            doc = "List of packages to not download",
+            default = _EXCLUDE_BASE_PKGS, 
+        ),
+    }
+)
+
+def ubuntu_base_rule_impl(ctx):
+    outfile = ctx.actions.declare_file("%s.tar" % ctx.attr.name)
+    args = ctx.actions.args()
+    args.add(ctx.attr.arch)
+    args.add(",".join(ctx.attr.components))
+    args.add(ctx.attr.distro)
+    args.add(ctx.attr.mirror)
+    args.add(outfile)
+    args.add_all(ctx.attr.pkgs)
+
+    ctx.actions.run(
+        executable = ctx.executable.bootstrap_script,
+        outputs = [outfile],
+        arguments = [args],
+    )
+
+    return [
+        DefaultInfo(files = depset([outfile])),
+    ]
+
+
+ubuntu_base_rule = rule(
+    implementation = ubuntu_base_rule_impl,
+    attrs = bootstrap_attrs | {
+        "pkgs": attr.string_list(
+            doc = "List of packages to include in the base image",
+            default = ["curl", "wget", "software-properties-common"]
```

Comment:
Yes, because in the `preinstall_ubuntu.sh` script, sometimes I need to curl, wget, and apt-add-repository some third-party repos like docker, gcloud cli, etc.

Reply:
makes sense 👍🏼